### PR TITLE
DiagGGN : new test suite

### PR DESCRIPTION
**Pull request by sbharadwajj**
New test suite for `diag_ggn`
Note: test for `diag_ggn_mc` is skipped

### CODE REVIEW

**Comment by sbharadwajj** [comment]
rechecked and the test passes locally

File: test/extensions/implementation/backpack.py [hunk 1]
```diff
@@ -39,6 +39,20 @@ def variance(self):
...
         return variances
 
+    def diag_ggn(self):
+        with backpack(new_ext.DiagGGNExact()):
+            _, _, loss = self.problem.forward_pass()
+            loss.backward()
+            diag_ggn = [p.diag_ggn_exact for p in self.problem.model.parameters()]
+        return diag_ggn
+
+    def diag_ggn_mc(self):
```
Comment (f-dangel) [inline test/extensions/implementation/backpack.py:49]: Please include the option to use the `mc_samples` argument of `DiagGGNMC`

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        backpack_res_mc_avg.append(torch.zeros_like(param_res))
+
+    mc_samples = 500
+    for _ in range(mc_samples):
+        backpack_diagggn_mc = BackpackExtensions(problem).diag_ggn_mc()
+        for i, param_res in enumerate(backpack_diagggn_mc):
+            backpack_res_mc_avg[i] += param_res
+
+    for i in range(len(backpack_res_mc_avg)):
+        backpack_res_mc_avg[i] /= mc_samples
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:53]: With the above change, this should become a `backpack_res_mc_avg = BackpackExtensions(problem).diag_ggn_mc(mc_samples)`

File: test/extensions/implementation/base.py [hunk 3]
```diff
@@ -20,6 +20,14 @@ def variance(self):
         """Variance of Individual gradients"""
         raise NotImplementedError
 
+    def diag_ggn(self):
+        """Diagonal of Gauss Newton"""
+        raise NotImplementedError
+
+    def diag_ggn_mc(self):
```
Comment (f-dangel) [inline test/extensions/implementation/base.py:27]: Missing argument `mc_samples`

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py (hunk 2 above)
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py]: still open

Comment (sbharadwajj) [reply test/extensions/secondorder/diag_ggn/test_diag_ggn.py]: Sorry, theres some commit error

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    """
+    problem.set_up()
+    torch.manual_seed(0)
+
+    backpack_res = BackpackExtensions(problem).diag_ggn()
+
+    mc_samples = 10000
+    backpack_res_mc_avg = BackpackExtensions(problem).diag_ggn_mc(mc_samples)
+
+    check_sizes_and_values(backpack_res, backpack_res_mc_avg, atol=1e-2, rtol=1e-2)
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:68]: `atol` and `rtol` should be smaller than in `test_diag_ggn_mc_light` since 10x more samples are being used

File: .travis.yml [hunk 5]
```diff
@@ -10,7 +10,7 @@ install:
 cache:
 - pip
 script:
-- pytest -vx --cov=backpack/ .
+- if [ $TRAVIS_BRANCH = "master" || $TRAVIS_BRANCH = "development" ||  $TRAVIS_BRANCH = "release" ]; then pytest -vx --runslow --cov=backpack/ .; else pytest -vx --cov=backpack/ .; fi
```
Comment (f-dangel) [inline .travis.yml:13]: What is `--runslow` ?

Comment (sbharadwajj) [reply .travis.yml]: fixed

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py (hunk 4 above)
Comment (sbharadwajj) [reply test/extensions/secondorder/diag_ggn/test_diag_ggn.py]: verified if the test passes for smaller tolerance. `1e-3` requires `1000000` samples

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+@pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
+def test_diag_ggn_mc_light(problem):
+    """Test the MC approximation of Diagonal of Gauss-Newton
+        with few mc_samples (light version)
+
+    Args:
+        problem (ExtensionsTestProblem): Problem for extension test.
+    """
+    problem.set_up()
+    torch.manual_seed(0)
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:40]: `problem.set_up` already sets a seed. Please check, and if so, remove this line.

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+@pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
+def test_diag_ggn_mc(problem):
+    """Test the MC approximation of Diagonal of Gauss-Newton
+       with more samples (slow version)
+
+    Args:
+        problem (ExtensionsTestProblem): Problem for extension test.
+    """
+    problem.set_up()
+    torch.manual_seed(0)
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:61]: `problem.set_up` already sets a seed. Please check, and if so, remove this line.

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+
+@pytest.mark.montecarlo
+@pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
+def test_diag_ggn_mc(problem):
+    """Test the MC approximation of Diagonal of Gauss-Newton
+       with more samples (slow version)
+
+    Note: This function runs for 100,000 mc_samples. In case this function crashes,
+    alternate approach is to loop 10 times over 10,000 samples & then average them.
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:56]: Move above of `mc_samples = 100000` such that it's easier to find

Comment (sbharadwajj) [reply test/extensions/secondorder/diag_ggn/test_diag_ggn.py:56]: I dont exactly understand
